src/object/obj_tx.c

```diff
@@ -863,21 +1532,22 @@ dc_tx_commit(tse_task_t *task)
 	if (tx->tx_status == TX_COMMITTED)
 		D_GOTO(out_tx, rc = 0);
 
-	if (tx->tx_status == TX_COMMITTING) {
-		/* FIXME: Before support compounded RPC, the retry update/punch
-		 *	  RPC will hit TX_COMMITTING status TX, that is normal.
-		 */
-		D_ASSERT(tx->tx_write_cnt != 0);
-	} else if (tx->tx_status != TX_OPEN) {
+	if (tx->tx_status == TX_COMMITTING)
+		D_GOTO(out_tx, rc = -DER_INPROGRESS);
+
+	if (tx->tx_status != TX_OPEN && !(tx->tx_status == TX_FAILED &&
+	    tx->tx_retry && args->flags & DTF_RETRY_COMMIT)) {
```

Comment:
Alignment? I was tricked by it for a moment.

Reply:
OK, fixed.